### PR TITLE
Feature/4ormoredots

### PR DESCRIPTION
**Pull request by StabbyCutyou**
We ran into an issue where we had version strings with more specificity than 3 dots - We had to hack around it for the time being, but we'd ultimately appreciate if the library could natively support this.

I took a crack at changing the Regex - It now supports an unlimited number of dots (because no matter what number I pick, someone else will need it bumped up by 1 some day), and altered a few things to support that (removing the cap on the segments slice),  and added some comments around why some code still refers to 3 places (such as where you handle the segments being less than 3 in number, to support the default case of MAJOR.MINOR.PATCH).

I also had to update the way the pessimistic comparator worked, to cover the additional edge cases present in the ability to support a greater specificity, as well as problems stemming from a jagged or uneven specificity between constraints and versions that I wasn't sure was explicitly being covered properly.

I also added test cases to cover this.

And I fixed an annoying nag from the go tools about replacing +=1 with ++, along with one about a syntax change to a range iterator.

### CODE REVIEW

File: version.go [hunk 1]
```diff
@@ -107,20 +110,30 @@ func (v *Version) Compare(other *Version) int {
...
+			break
 		} else {
-			return 1
+			matched = 1
+			break
 		}
 	}
 
-	panic("should not be reached")
+	return matched
```
Comment (StabbyCutyou) [inline version.go]: I got rid of this for 2 reasons:

1. I don't want to be around when this edge case managed to happen :)
2. You really should never panic in a client library

**Comment by StabbyCutyou** [comment]
I also added a lot of heavy comments around any of the code I touched, for the hell of it.

**Comment by deanxorr** [comment]
This is awesome @StabbyCutyou, would love to use this when it gets accepted.

File: version.go [hunk 2]
```diff
@@ -107,20 +110,30 @@ func (v *Version) Compare(other *Version) int {
...
+	// Because a constraint could have more/less specificity than the version it's
+	// checking, we need to account for a lopsided or jagged comparison
+	var matched int
+	for i, lhs := range segmentsSelf {
+		if i > len(segmentsOther)-1 {
+			// The constraint only had X places, but the version we're checking had
+			// X+1. This means our version is greater than the constraint version we're
+			// checking , ex: 1.2.3.4 > 1.2.3
+			matched = 1
+			break
```
Comment (StabbyCutyou) [inline version.go]: There may be a bug here, if the "excess" segments are all 0s, so i'll add a case in to test / handle this before merging.

**Comment by StabbyCutyou** [comment]
Any feedback on if this is something you guys are interested in merging in?

**Comment by thanasisk** [comment]
+1 for fixing this

**Comment by StabbyCutyou** [comment]
Thanks. Would love to get it merged in.

**Comment by zuBux** [comment]
:+1:  from me as well. Judging from the PR description, it will solve a few problems for me as well :)

**Comment by mitchellh** [comment]
Reviewing this now but the idea is fine to me!

**Comment by mitchellh** [comment]
LGTM. Merging it in.

**Comment by StabbyCutyou** [comment]
Thanks!

**Pull request merged**